From cf729cdbeeec35fcfce367084d08bf0bdc88b02d Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Wed, 29 May 2024 19:37:21 +0300
Subject: [PATCH] add duress password feature

Ported from GrapheneOS 2025061900 (886fc0f85b57); ILockSettings hunk fuzzed past LineageOS getLockPatternSize().

Change-Id: I937f9fef89fe1e4c5c131a3c022076b3bbff7cf6
---
 .../internal/widget/ILockSettings.aidl        |   2 +
 .../internal/widget/LockPatternUtils.java     |  34 ++++++
 .../server/locksettings/DuressCredential.java |  66 +++++++++++
 .../locksettings/DuressCredentials.java       | 104 +++++++++++++++++
 .../locksettings/DuressPasswordHelper.java    | 109 ++++++++++++++++++
 .../server/locksettings/DuressWipe.java       |  77 +++++++++++++
 .../locksettings/LockSettingsService.java     |  48 ++++++++
 .../recoverysystem/RecoverySystemService.java |   2 +-
 .../LockSettingsServiceTestable.java          |   7 ++
 9 files changed, 448 insertions(+), 1 deletion(-)
 create mode 100644 services/core/java/com/android/server/locksettings/DuressCredential.java
 create mode 100644 services/core/java/com/android/server/locksettings/DuressCredentials.java
 create mode 100644 services/core/java/com/android/server/locksettings/DuressPasswordHelper.java
 create mode 100644 services/core/java/com/android/server/locksettings/DuressWipe.java

diff --git a/core/java/com/android/internal/widget/ILockSettings.aidl b/core/java/com/android/internal/widget/ILockSettings.aidl
index a19dc669f..701fb859b 100644
--- a/core/java/com/android/internal/widget/ILockSettings.aidl
+++ b/core/java/com/android/internal/widget/ILockSettings.aidl
@@ -110,5 +110,7 @@ interface ILockSettings {
     boolean isWeakEscrowTokenValid(long handle, in byte[] token, int userId);
     void unlockUserKeyIfUnsecured(int userId);
     boolean writeRepairModeCredential(int userId);
+    void setDuressCredentials(in LockscreenCredential ownerCredential, in LockscreenCredential duressPin, in LockscreenCredential duressPassword);
+    boolean hasDuressCredentials(in LockscreenCredential ownerCredential);
     byte getLockPatternSize(int userId);
 }
diff --git a/core/java/com/android/internal/widget/LockPatternUtils.java b/core/java/com/android/internal/widget/LockPatternUtils.java
index b499ea20f..9ef9b2cc8 100644
--- a/core/java/com/android/internal/widget/LockPatternUtils.java
+++ b/core/java/com/android/internal/widget/LockPatternUtils.java
@@ -2090,4 +2090,38 @@ public class LockPatternUtils {
             throw e.rethrowFromSystemServer();
         }
     }
+
+    public static List<PasswordValidationError> validateDuressCredential(
+            @NonNull LockscreenCredential credential) {
+        var metrics = new PasswordMetrics(credential.getType());
+        metrics.length = 4;
+        return PasswordMetrics.validateCredential(metrics, DevicePolicyManager.PASSWORD_COMPLEXITY_LOW, credential);
+    }
+
+    public void setDuressCredentials(@NonNull LockscreenCredential ownerCredential,
+                                     @NonNull LockscreenCredential duressPin,
+                                     @NonNull LockscreenCredential duressPassword) {
+        try {
+            getLockSettings().setDuressCredentials(ownerCredential, duressPin, duressPassword);
+        } catch (RemoteException e) {
+            throw e.rethrowFromSystemServer();
+        }
+    }
+
+    public void deleteDuressCredentials(@NonNull LockscreenCredential ownerCredential) {
+        var noCredential = LockscreenCredential.createNone();
+        try {
+            getLockSettings().setDuressCredentials(ownerCredential, noCredential, noCredential);
+        } catch (RemoteException e) {
+            throw e.rethrowFromSystemServer();
+        }
+    }
+
+    public boolean hasDuressCredentials(@NonNull LockscreenCredential ownerCredential) {
+        try {
+            return getLockSettings().hasDuressCredentials(ownerCredential);
+        } catch (RemoteException e) {
+            throw e.rethrowFromSystemServer();
+        }
+    }
 }
diff --git a/services/core/java/com/android/server/locksettings/DuressCredential.java b/services/core/java/com/android/server/locksettings/DuressCredential.java
new file mode 100644
index 000000000..96781687a
--- /dev/null
+++ b/services/core/java/com/android/server/locksettings/DuressCredential.java
@@ -0,0 +1,66 @@
+package com.android.server.locksettings;
+
+import com.android.internal.widget.LockPatternUtils;
+import com.android.internal.widget.LockscreenCredential;
+import com.android.internal.widget.PasswordValidationError;
+import com.android.server.locksettings.SyntheticPasswordManager.PasswordData;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+class DuressCredential {
+    private final PasswordData salt;
+    private final byte[] hashedCredential;
+
+    DuressCredential(PasswordData salt, byte[] hashedCredential) {
+        this.salt = salt;
+        this.hashedCredential = hashedCredential;
+    }
+
+    static DuressCredential create(LockscreenCredential credential, SyntheticPasswordManager spm) {
+        PasswordData salt = PasswordData.create(credential.getType(),
+                // no need to leak PIN length here, it's used only for PIN auto-confirm
+                LockPatternUtils.PIN_LENGTH_UNAVAILABLE);
+        byte[] hashedCredential = spm.stretchLskf(credential, salt);
+        return new DuressCredential(salt, hashedCredential);
+    }
+
+    boolean verify(SyntheticPasswordManager spm, LockscreenCredential credential) {
+        return Arrays.equals(hashedCredential, spm.stretchLskf(credential, salt));
+    }
+
+    int getType() {
+        return salt.credentialType;
+    }
+
+    void serialize(DataOutputStream s) throws IOException {
+        byte[] saltBytes = salt.toBytes();
+        s.writeInt(saltBytes.length);
+        s.write(saltBytes);
+        s.writeInt(hashedCredential.length);
+        s.write(hashedCredential);
+    }
+
+    static DuressCredential deserialize(DataInputStream s) throws IOException {
+        PasswordData salt = PasswordData.fromBytes(s.readNBytes(s.readInt()));
+        byte[] hashedCredential = s.readNBytes(s.readInt());
+        return new DuressCredential(salt, hashedCredential);
+    }
+
+    static void validate(LockscreenCredential credential, int expectedType) {
+        int type = credential.getType();
+        if (type != expectedType) {
+            throw new IllegalArgumentException("type mismatch: expected " + expectedType + ", got " + type);
+        }
+
+        List<PasswordValidationError> validationErrors =
+                LockPatternUtils.validateDuressCredential(credential);
+        if (!validationErrors.isEmpty()) {
+            throw new IllegalArgumentException("validation failed: " +
+                    Arrays.toString(validationErrors.toArray()));
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/locksettings/DuressCredentials.java b/services/core/java/com/android/server/locksettings/DuressCredentials.java
new file mode 100644
index 000000000..7a2312ef5
--- /dev/null
+++ b/services/core/java/com/android/server/locksettings/DuressCredentials.java
@@ -0,0 +1,104 @@
+package com.android.server.locksettings;
+
+import android.annotation.Nullable;
+import android.database.sqlite.SQLiteDiskIOException;
+import android.os.UserHandle;
+
+import com.android.internal.widget.LockscreenCredential;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+import libcore.util.HexEncoding;
+
+import static com.android.internal.widget.LockPatternUtils.CREDENTIAL_TYPE_PASSWORD;
+import static com.android.internal.widget.LockPatternUtils.CREDENTIAL_TYPE_PIN;
+
+class DuressCredentials {
+    final DuressCredential pin;
+    final DuressCredential password;
+
+    DuressCredentials(DuressCredential pin, DuressCredential password) {
+        if (pin.getType() != CREDENTIAL_TYPE_PIN) {
+            throw new IllegalArgumentException();
+        }
+        this.pin = pin;
+        if (password.getType() != CREDENTIAL_TYPE_PASSWORD) {
+            throw new IllegalArgumentException();
+        }
+        this.password = password;
+    }
+
+    static DuressCredentials create(SyntheticPasswordManager spm,
+                                    LockscreenCredential pin, LockscreenCredential password) {
+        return new DuressCredentials(DuressCredential.create(pin, spm),
+                DuressCredential.create(password, spm));
+    }
+
+    private static final String LOCK_SETTINGS_STORAGE_KEY = "duress_credentials";
+    private static final int LOCK_SETTINGS_STORAGE_USER_ID = UserHandle.USER_SYSTEM;
+
+    void save(LockSettingsStorage lss) {
+        lss.setString(LOCK_SETTINGS_STORAGE_KEY, serialize(), LOCK_SETTINGS_STORAGE_USER_ID);
+    }
+
+    @Nullable
+    static DuressCredentials maybeGet(LockSettingsStorage lss) {
+        String s = lss.getString(LOCK_SETTINGS_STORAGE_KEY, null, LOCK_SETTINGS_STORAGE_USER_ID);
+        if (s == null) {
+            return null;
+        }
+        return deserialize(s);
+    }
+
+    static void delete(LockSettingsStorage lss) {
+        lss.removeKey(LOCK_SETTINGS_STORAGE_KEY, LOCK_SETTINGS_STORAGE_USER_ID);
+    }
+
+    DuressCredential get(int type) {
+        return switch (type) {
+            case CREDENTIAL_TYPE_PIN -> pin;
+            case CREDENTIAL_TYPE_PASSWORD -> password;
+            default -> throw new IllegalArgumentException(Integer.toString(type));
+        };
+    }
+
+    private static final int VERSION = 0;
+
+    private String serialize() {
+        var bos = new ByteArrayOutputStream(1000);
+        var s = new DataOutputStream(bos);
+        try {
+            s.writeByte(VERSION);
+            pin.serialize(s);
+            password.serialize(s);
+        } catch (IOException e) {
+            // ByteArrayOutputStream never throws IOException
+            throw new IllegalStateException(e);
+        }
+
+        return HexEncoding.encodeToString(bos.toByteArray());
+    }
+
+    private static DuressCredentials deserialize(String str) {
+        var s = new DataInputStream(new ByteArrayInputStream(HexEncoding.decode(str)));
+        try {
+            int version = s.readByte();
+            if (version > VERSION) {
+                throw new IllegalArgumentException(str);
+            }
+            var pin = DuressCredential.deserialize(s);
+            var password = DuressCredential.deserialize(s);
+            if (s.available() != 0) {
+                throw new IllegalArgumentException(str);
+            }
+            return new DuressCredentials(pin, password);
+        } catch (IOException e) {
+            // ByteArrayInputStream never throws IOException
+            throw new IllegalStateException(e);
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/locksettings/DuressPasswordHelper.java b/services/core/java/com/android/server/locksettings/DuressPasswordHelper.java
new file mode 100644
index 000000000..d57682c4b
--- /dev/null
+++ b/services/core/java/com/android/server/locksettings/DuressPasswordHelper.java
@@ -0,0 +1,109 @@
+package com.android.server.locksettings;
+
+import android.annotation.Nullable;
+import android.os.HandlerThread;
+import android.os.UserHandle;
+import android.util.Slog;
+
+import com.android.internal.widget.LockscreenCredential;
+import com.android.internal.widget.VerifyCredentialResponse;
+
+import java.util.Objects;
+import java.util.UUID;
+
+import static com.android.internal.widget.LockDomain.Primary;
+import static com.android.internal.widget.LockPatternUtils.CREDENTIAL_TYPE_NONE;
+import static com.android.internal.widget.LockPatternUtils.CREDENTIAL_TYPE_PASSWORD;
+import static com.android.internal.widget.LockPatternUtils.CREDENTIAL_TYPE_PIN;
+
+public class DuressPasswordHelper {
+    static final String TAG = DuressPasswordHelper.class.getSimpleName();
+
+    private final LockSettingsService lockSettingsService;
+    private final LockSettingsStorage lockSettingsStorage;
+    private final SyntheticPasswordManager spManager;
+    private final HandlerThread backgroundThread;
+
+    DuressPasswordHelper(LockSettingsService lockSettingsService,
+            LockSettingsStorage lockSettingsStorage, SyntheticPasswordManager spManager) {
+        var bgThread = new HandlerThread(UUID.randomUUID().toString());
+        bgThread.start();
+        this.backgroundThread = bgThread;
+        this.lockSettingsService = lockSettingsService;
+        this.lockSettingsStorage = lockSettingsStorage;
+        this.spManager = spManager;
+    }
+
+    protected void onVerifyCredentialResult(@Nullable VerifyCredentialResponse res, @Nullable LockscreenCredential credential) {
+        if (res != null && res.getResponseCode() == VerifyCredentialResponse.RESPONSE_OK) {
+            return;
+        }
+
+        if (credential == null) {
+            return;
+        }
+
+        // credential verification is slow, don't block the current (usually binder) thread
+        backgroundThread.getThreadHandler().post(() -> {
+            if (isDuressCredential(credential)) {
+                DuressWipe.run(lockSettingsService.getContext());
+            }
+        });
+    }
+
+    private void checkOwnerCredential(LockscreenCredential ownerCredential) {
+        int userId = UserHandle.USER_SYSTEM;
+
+        if (lockSettingsService.getCredentialType(userId) == CREDENTIAL_TYPE_NONE) {
+            if (!ownerCredential.isNone()) {
+                throw new IllegalArgumentException("!ownerCredential.isNone()");
+            }
+        } else {
+            VerifyCredentialResponse response = lockSettingsService.checkCredential(ownerCredential,
+                    Primary, userId, null);
+
+            if (response.getResponseCode() != VerifyCredentialResponse.RESPONSE_OK) {
+                throw new SecurityException("owner credential verification failed; " + response);
+            }
+        }
+    }
+
+    protected void setDuressCredentials(LockscreenCredential ownerCredential,
+                                 LockscreenCredential pin, LockscreenCredential password) {
+        Objects.requireNonNull(ownerCredential, "ownerCredential");
+        Objects.requireNonNull(pin, "pin");
+        Objects.requireNonNull(password, "password");
+
+        checkOwnerCredential(ownerCredential);
+
+        if (pin.isNone() && password.isNone()) {
+            // exception handling is delegated to the caller
+            DuressCredentials.delete(lockSettingsStorage);
+            Slog.d(TAG, "deleted duress credentials");
+            return;
+        }
+
+        DuressCredential.validate(pin, CREDENTIAL_TYPE_PIN);
+        DuressCredential.validate(password, CREDENTIAL_TYPE_PASSWORD);
+
+        // exception handling is delegated to the caller
+        DuressCredentials.create(spManager, pin, password).save(lockSettingsStorage);
+    }
+
+    protected boolean hasDuressCredentials(LockscreenCredential ownerCredential) {
+        checkOwnerCredential(ownerCredential);
+        return DuressCredentials.maybeGet(lockSettingsStorage) != null;
+    }
+
+    private boolean isDuressCredential(LockscreenCredential credential) {
+        int credentialType = credential.getType();
+        switch (credentialType) {
+            case CREDENTIAL_TYPE_PIN:
+            case CREDENTIAL_TYPE_PASSWORD:
+                DuressCredentials dc = DuressCredentials.maybeGet(lockSettingsStorage);
+                return dc != null && dc.get(credentialType).verify(spManager, credential);
+            default:
+                return false;
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/locksettings/DuressWipe.java b/services/core/java/com/android/server/locksettings/DuressWipe.java
new file mode 100644
index 000000000..cdc63df57
--- /dev/null
+++ b/services/core/java/com/android/server/locksettings/DuressWipe.java
@@ -0,0 +1,77 @@
+package com.android.server.locksettings;
+
+import android.annotation.UptimeMillisLong;
+import android.content.Context;
+import android.os.SystemClock;
+import android.telephony.euicc.EuiccCardManager;
+import android.util.Slog;
+
+import com.android.internal.telephony.euicc.EuiccWipe;
+import com.android.server.power.PowerManagerService;
+import com.android.server.recoverysystem.RecoverySystemService;
+
+public class DuressWipe {
+    static final String TAG = DuressWipe.class.getSimpleName();
+
+    // used only for testing, guarded by owner credential
+    public static boolean sleep5sBeforePoweroff;
+
+    static void run(Context context) {
+        Slog.d(TAG, "start");
+
+        EuiccWipeThread euiccWipeThread = EuiccWipeThread.start(context);
+
+        Slog.d(TAG, "calling deleteSecrets");
+        // deleteSecrets() calls AndroidKeyStoreMaintenance.deleteAllKeys(), which deletes all
+        // KeyMint keys, including the storage encryption keys
+        RecoverySystemService.deleteSecrets();
+        Slog.d(TAG, "deleteSecrets returned");
+
+        euiccWipeThread.await(3000);
+        Slog.d(TAG, "finished waiting for euiccWipeThread");
+
+        if (sleep5sBeforePoweroff) {
+            SystemClock.sleep(5000);
+        }
+
+        PowerManagerService.lowLevelShutdown(null);
+    }
+
+    static class EuiccWipeThread extends Thread {
+        private final Context context;
+        @UptimeMillisLong private long startTime;
+
+        private EuiccWipeThread(Context context) {
+            this.context = context;
+        }
+
+        static EuiccWipeThread start(Context ctx) {
+            var t = new EuiccWipeThread(ctx);
+            t.start();
+            t.startTime = SystemClock.uptimeMillis();
+            return t;
+        }
+
+        @Override
+        public void run() {
+            try {
+                EuiccWipe.run(context, EuiccCardManager.RESET_FLAG_IS_FOR_DURESS_WIPE);
+            } catch (Throwable e) {
+                Slog.e(TAG, "", e);
+            }
+        }
+
+        void await(long timeoutMs) {
+            long remaining = timeoutMs - (SystemClock.uptimeMillis() - startTime);
+            if (remaining <= 0) {
+                return;
+            }
+            try {
+                join(remaining);
+            } catch (InterruptedException e) {
+                // should never happen, but don't rethrow it just in case
+                Slog.e(TAG, "", e);
+            }
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/locksettings/LockSettingsService.java b/services/core/java/com/android/server/locksettings/LockSettingsService.java
index 4682a37b6..fed71536f 100644
--- a/services/core/java/com/android/server/locksettings/LockSettingsService.java
+++ b/services/core/java/com/android/server/locksettings/LockSettingsService.java
@@ -94,6 +94,7 @@ import android.os.Bundle;
 import android.os.Handler;
 import android.os.IBinder;
 import android.os.IProgressListener;
+import android.os.ParcelableException;
 import android.os.Process;
 import android.os.RemoteException;
 import android.os.ResultReceiver;
@@ -607,6 +608,13 @@ public class LockSettingsService extends ILockSettings.Stub {
                     getHandler(getServiceThread()), getUserManagerInternal());
         }
 
+        public DuressPasswordHelper getDuressPasswordHelper(LockSettingsService lockSettingsService,
+                LockSettingsStorage lockSettingsStorage,
+                SyntheticPasswordManager syntheticPasswordManager) {
+            return new DuressPasswordHelper(lockSettingsService, lockSettingsStorage,
+                    syntheticPasswordManager);
+        }
+
         public int binderGetCallingUid() {
             return Binder.getCallingUid();
         }
@@ -703,6 +711,8 @@ public class LockSettingsService extends ILockSettings.Stub {
                 mStorage);
 
         LocalServices.addService(LockSettingsInternal.class, new LocalService());
+
+        duressPasswordHelper = injector.getDuressPasswordHelper(this, mStorage, mSpManager);
     }
 
     private void updateActivatedEncryptionNotifications(String reason) {
@@ -2427,6 +2437,18 @@ public class LockSettingsService extends ILockSettings.Stub {
     private VerifyCredentialResponse doVerifyCredential(LockscreenCredential credential,
             int userId, ICheckCredentialProgressCallback progressCallback,
             @LockPatternUtils.VerifyFlag int flags) {
+        VerifyCredentialResponse res = null;
+        try {
+            res = doVerifyCredentialInner(credential, userId, progressCallback, flags);
+            return res;
+        } finally {
+            duressPasswordHelper.onVerifyCredentialResult(res, credential);
+        }
+    }
+
+    private VerifyCredentialResponse doVerifyCredentialInner(LockscreenCredential credential,
+            int userId, ICheckCredentialProgressCallback progressCallback,
+            @LockPatternUtils.VerifyFlag int flags) {
         if (credential == null || credential.isNone()) {
             throw new IllegalArgumentException("Credential can't be null or empty");
         }
@@ -3876,4 +3898,30 @@ public class LockSettingsService extends ILockSettings.Stub {
             onCredentialVerified(sp, loadPasswordMetrics(sp, userId), userId);
         }
     }
+
+    private final DuressPasswordHelper duressPasswordHelper;
+
+    @Override
+    public void setDuressCredentials(LockscreenCredential ownerCredential,
+            LockscreenCredential pin, LockscreenCredential password) {
+        checkWritePermission();
+
+        try {
+            duressPasswordHelper.setDuressCredentials(ownerCredential, pin, password);
+        } catch (Throwable e) {
+            throw new ParcelableException(e);
+        } finally {
+            scheduleGc();
+        }
+    }
+
+    @Override
+    public boolean hasDuressCredentials(LockscreenCredential ownerCredential) {
+        checkPasswordHavePermission();
+        return duressPasswordHelper.hasDuressCredentials(ownerCredential);
+    }
+
+    Context getContext() {
+        return mContext;
+    }
 }
diff --git a/services/core/java/com/android/server/recoverysystem/RecoverySystemService.java b/services/core/java/com/android/server/recoverysystem/RecoverySystemService.java
index 8042165f1..2078c3fcd 100644
--- a/services/core/java/com/android/server/recoverysystem/RecoverySystemService.java
+++ b/services/core/java/com/android/server/recoverysystem/RecoverySystemService.java
@@ -551,7 +551,7 @@ public class RecoverySystemService extends IRecoverySystem.Stub implements Reboo
         }
     }
 
-    private static void deleteSecrets() {
+    public static void deleteSecrets() {
         Slogf.w(TAG, "deleteSecrets");
         try {
             AndroidKeyStoreMaintenance.deleteAllKeys();
diff --git a/services/tests/servicestests/src/com/android/server/locksettings/LockSettingsServiceTestable.java b/services/tests/servicestests/src/com/android/server/locksettings/LockSettingsServiceTestable.java
index abd39b0bb..c486f8fa1 100644
--- a/services/tests/servicestests/src/com/android/server/locksettings/LockSettingsServiceTestable.java
+++ b/services/tests/servicestests/src/com/android/server/locksettings/LockSettingsServiceTestable.java
@@ -154,6 +154,13 @@ public class LockSettingsServiceTestable extends LockSettingsService {
         public boolean isMainUserPermanentAdmin() {
             return mIsMainUserPermanentAdmin;
         }
+
+        @Override
+        public DuressPasswordHelper getDuressPasswordHelper(LockSettingsService lockSettingsService,
+                LockSettingsStorage lockSettingsStorage,
+                SyntheticPasswordManager syntheticPasswordManager) {
+            return mock(DuressPasswordHelper.class);
+        }
     }
 
     protected LockSettingsServiceTestable(
-- 
2.43.0

